### DISCUSSION TITLE
"Linked" preview

### DISCUSSION THREAD

**Discussion by edrex** [srid/ema#59] [Ideas]
I'm using emanote with Obsidian (and Vim, Emacs, etc) and have been thinking about ways of enabling side-by-side preview without requiring explicit navigation when switching pages. There are plenty of possible approaches:

 - Embed a webview in editor (eg obsidian) replacing the stock preview with emanote, possibly launching it
 - Various "edit this page" schemes using eg `<link rel=edit...` and either a browser extension or some browser-based editor.

I'm more interested in the second class of solutions, but there is another option that I wanted to put up for discussion: in hot reload (server) mode, Ema is watching source files for changes, updating its model and sending reload commands down the websocket. What if it also learned to send redirect commands to the slug for whatever file changed most recently? Some aggregation would be necessary for bulk updates (maybe redirect to the first change event right away, then wait a bit and redirect to the last one, so max 2 redirects per batch). This is just one idea, but I like that it completely decouples from the editor and uses filesystem updates for signaling (i mean, monitoring FS updates is super difficult and edge-casey, but since ema is already doing it, it probably wouldn't add too much overhead or code complexity, i guess)

**Comment by edrex** [comment]
There would need to be a way to switch the redirects off, probably with a flag, and maybe also opt-in via a `#followedits` fragment ID that the client side of the websocket could respect (or maybe a `?followedits` get param could enable the behavior on the server without needing to add a command line flag, and without possibly interfering with some user's client side fragment routing )

**Comment by edrex** [comment]
related: https://github.com/srid/emanote/discussions/121

**Comment by srid** [comment]
> What if it also learned to send redirect commands to the slug for whatever file changed most recently?

Yes, I too thought of this idea in the context of Emanote. We should do it for sure. Perhaps Ema can provide a "frontend API" that talks to the web browser through websocket, with a single API entry in it for now called "switchRoute". Then Emanote can call `switchRoute` for every latest changed file (with some aggregation delay) to "sync" the view with your editor.

We should also consider this in terms of #55 (and vice-versa) as well as https://github.com/srid/emanote/discussions/84.

But maybe for initial rough version we can just expose `switchRoute` as an experimental API.

**Reply by edrex** [reply]
I see the current websocket endpoint is pretty simple: path up (and a second path up to change route), html fragment down.

Would you rather change the current impl to support some message type specifier (at least down, but might as well be both ways) anticipating future features, or add another endpoint for the server->client switchRoute messages?

Alternatively, the current endpoint could be changed so the server always sends a route before the html fragment, and the client updated to support setting the route from the server. This seems like the minimal change needed. How about it?

**Reply by srid** [reply]
We have `switchRoute` from JS side, https://github.com/srid/ema/issues/69 - what's left is to make `switchRoute` from Haskell side?

@lucasvreis asked for this here: https://github.com/srid/ema/pull/70#issuecomment-1026078962

**Comment by srid** [comment]
> Embed a webview in editor (eg obsidian) replacing the stock preview with emanote, possibly launching it

Vivaldi has [split view](https://help.vivaldi.com/tutorials/create-a-split-screen-view-using-tab-selection/), which in conjunction with something like running vim over [gotty](https://github.com/yudai/gotty), can be interesting to achieve this workflow *other way* (ie. editor in web browser, rather than web browser in editor).

![image](https://user-images.githubusercontent.com/3998/137598254-76a5b961-60bd-46ea-a6b0-c5233e1d2e8e.png)

**Comment by srid** [comment]
https://github.com/srid/ema/issues/69

**Comment by srid** [comment]
@edrex See https://github.com/srid/ema/pull/152
